src/models/Call.ts

```diff
@@ -537,3 +552,253 @@ export class JitsiCall extends Call {
+        const devices = Array.isArray(call?.["m.devices"]) ? call["m.devices"] : [];
+        const newDevices = fn(devices.map(d => d["m.device_id"]));
+
+        if (newDevices) {
+            const content: ElementCallMemberContent = {
+                "m.expires_ts": Date.now() + ElementCall.STUCK_DEVICE_TIMEOUT_MS,
+                "m.calls": [
+                    {
+                        "m.call_id": this.groupCall.getStateKey(),
+                        "m.devices": newDevices.map(d => ({ "m.device_id": d })),
+                    },
+                    ...calls.filter(c => c["m.call_id"] !== this.groupCall.getStateKey()),
+                ],
+            };
+
+            await this.client.sendStateEvent(
+                this.roomId, ElementCall.MEMBER_EVENT_TYPE.name, content, this.client.getUserId()!,
+            );
+        }
+    }
+
+    public async clean(): Promise<void> {
+        const now = Date.now();
+        const { devices: myDevices } = await this.client.getDevices();
+        const deviceMap = new Map<string, IMyDevice>(myDevices.map(d => [d.device_id, d]));
+
+        // Clean up our member state by filtering out logged out devices,
+        // inactive devices, and our own device (if we're disconnected)
+        await this.updateDevices(devices => {
+            const newDevices = devices.filter(d => {
+                const device = deviceMap.get(d);
+                return device?.last_seen_ts
+                    && !(d === this.client.getDeviceId() && !this.connected)
+                    && (now - device.last_seen_ts) < ElementCall.STUCK_DEVICE_TIMEOUT_MS;
+            });
+
+            // Skip the update if the devices are unchanged
+            return newDevices.length === devices.length ? null : newDevices;
+        });
+    }
+
+    protected async performConnection(
+        audioInput: MediaDeviceInfo | null,
+        videoInput: MediaDeviceInfo | null,
+    ): Promise<void> {
+        try {
+            await this.messaging!.transport.send(ElementWidgetActions.JoinCall, {
+                audioInput: audioInput?.deviceId ?? null,
+                videoInput: videoInput?.deviceId ?? null,
+            });
+        } catch (e) {
+            throw new Error(`Failed to join call in room ${this.roomId}: ${e}`);
+        }
+
+        this.messaging!.on(`action:${ElementWidgetActions.HangupCall}`, this.onHangup);
+        this.room.on(RoomEvent.MyMembership, this.onMyMembership);
+        window.addEventListener("beforeunload", this.beforeUnload);
+    }
+
+    protected async performDisconnection(): Promise<void> {
+        try {
+            await this.messaging!.transport.send(ElementWidgetActions.HangupCall, {});
+        } catch (e) {
+            throw new Error(`Failed to hangup call in room ${this.roomId}: ${e}`);
+        }
+    }
+
+    public setDisconnected() {
+        this.messaging!.off(`action:${ElementWidgetActions.HangupCall}`, this.onHangup);
```

Comment:
Should `participantsExpirationTimer` be cleared here or in `destroy`?

Reply:
Good catch, I forgot about that